### ISSUE TITLE
Stuck on "Solving Environment"

### ISSUE THREAD

**Issue by AMElmore** [marykthompson/ribopop_probe_design#1]
I am rather new to bioinformatics, and I am trying to use this pipeline to design custom probes for Coccidioides immitis rRNA depletion. I have the ribopop_probe_design-master file on my desktop and have used terminal to try and set up the conda environment. I have miniconda installed and I am working out of the /ribopop_probe_design-master/envs/ directory. When I run the command "conda env create -f ribopop_probe_design.yaml, it returns "Collecting package metadata (repodata.json): done" and then "Solving environment: /" but then it gets stuck and does not return anything else. Do you have any idea why I might be running into this problem?

**Comment by marykthompson** [comment]
Hi, what OS are you using? I can try to get you an exact list of specifications that I used for MacOS or Linux that should be compatible with each other.

Other ideas:
- Make sure that you installed Miniconda for Python 3.
- Unfortunately, the 'solving environment' stage does often take a very long time. It is looking for versions of software that are all compatible with each other. If you wait long enough, it will probably spit out the incompatibilities and you could try to address those specifically.

**Comment by AMElmore** [comment]
Thank you for responding so soon. I'm using macOS Big Sur Version 11.2.1. 

I double checked and yes, I have miniconda for Python 3 installed. 

'Solving environment' is still running, but I'll let it keep going and hopefully it will spit out the incompatibilities like you said.

**Comment by marykthompson** [comment]
I'm attaching the yml file containing the exact specification that I'm running on my mac. Hopefully this will work for you -- let me know.
[ribopop_probe_design.txt](https://github.com/marykthompson/ribopop_probe_design/files/6085421/ribopop_probe_design.txt)

Github doesn't like the extension .yml, but it is just a text file. You can change the extension to .yml in case that makes a difference for conda.

You can install it with 'conda env create -f ribopop_probe_design.yml'

**Comment by AMElmore** [comment]
That worked, thank you! Unfortunately now I'm getting this error whenever I try to run the command 'snakemake --directory <my output directory>'

'Error: you need to specify the maximum number of CPU cores to be used at the same time with --cores.'

**Comment by marykthompson** [comment]
That should be simple to fix. In the more recent versions of snakemake you need to specify the cores argument. Just try '--cores 2' if you want it to use two, for example.

**Comment by AMElmore** [comment]
Ah thank you! I feel like I'm trying to run this out of the wrong directory though, now I keep getting the error 'Error: no Snakefile found, tried Snakefile, snakefile, workflow/Snakefile, workflow/snakefile.'

**Comment by marykthompson** [comment]
Sorry if this is not clear from the README.

Yes, you should run the pipeline from the directory where the Snakefile is. This will be the directory called ribopop_probe_design unless you renamed it after cloning or downloading the project.

For organizational purposes, I recommend putting your configuration files (i.e. your version of the ones shown here https://github.com/marykthompson/ribopop_probe_design/tree/master/example_outdir) in a separate directory. When you run snakemake with the --directory flag, you are telling it to use the pipeline with this directory.

Your command would be something like:
snakemake --directory <my_project_path>/Cimmitis_probes --cores 2

(where <my_project_path>/Cimmitis_probes is a directory containing a folder design_params/ with your csv files and config.yml)

**Comment by AMElmore** [comment]
Thank you so much for all your help, I'm sorry to keep bothering you. Moving them into their own directory worked, and I'm getting it to run, but it's getting stuck on line 12. I think I formatted my targets.csv file incorrectly, but I'm having trouble figuring out how to fix it. 
![Screen Shot 2021-03-04 at 6 46 28 PM](https://user-images.githubusercontent.com/46757230/110055310-fdc2f980-7d19-11eb-885e-31e23cad8c4f.png)

**Comment by marykthompson** [comment]
It looks OK, but can be hard to tell from a screenshot. If you want, you can post your param files here and I can try to see if it runs for me.

**Comment by AMElmore** [comment]
It looks like my params.csv file was the one that was formatted incorrectly - I fixed it and got a little further in the pipeline, but now I'm getting this error. I checked 'seqs_and_anns.csv' but it appears to be ok, I'll post them anyway though just in case I've missed something. 

[Archive.zip](https://github.com/marykthompson/ribopop_probe_design/files/6092192/Archive.zip)


Error in rule download_seqs:
    jobid: 8
    output: offtarget_filtering/Cimmitis/ncrna.fa, offtarget_filtering/Cimmitis/cdna.fa, offtarget_filtering/Cimmitis/ann.gtf, offtarget_filtering/Cimmitis/genome.fa

RuleException:
CalledProcessError in line 75 of /Users/alexiselmore/Desktop/ribopop_probe_design-master/Snakefile:
Command 'set -euo pipefail;  /Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/bin/python3.8 /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmpv6eaja0o.download_seqs.py' returned non-zero exit status 1.
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/Users/alexiselmore/Desktop/ribopop_probe_design-master/Snakefile", line 75, in __rule_download_seqs
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper

**Comment by marykthompson** [comment]
I think I may have spotted the problem. Your targets.csv file needs to have an ID or IDs matching the IDs in the non-coding RNA file from Ensembl. Otherwise, it has no way of retrieving the rRNA sequences. You have to go through that file and decide which IDs to include. If you look at the README and the example targets.csv file, I think you will see what I mean. 

One way of finding the IDs would be to use grep:
grep rRNA Coccidioides_immitis_rmscc_2394_gca_000149895.ASM14989v1.ncrna.fa

Here is the last line from that. 
>ENSRNAT049964321 ncrna supercontig:ASM14989v1:supercont1.8:264676:264795:-1 gene:ENSRNAG049964321 gene_biotype:rRNA transcript_biotype:rRNA gene_symbol:5S_rRNA description:5S ribosomal RNA

If you wanted to include this target, your corresponding line in the csv file would be something like:
5S,Cimmitis,ENSRNAT049964321 

You can include as many transcripts targeting the same gene as you like, just put them on another line, e.g. 
5S,Cimmitis,ENSRNAT049964327

If you'd rather use a different fasta file than the ncrna file from Ensemble to retrieve your rRNA sequences, that is also fine, but in this case you need to list the path to the file in the targets.csv file, e.g:

target,organism,ID,excluded_regions,target_subregions,file
18S,Hsap,NR_145820.1,"224-297,873-912",,/usr/people/myID/2.25.fishprobe_design/mammal_200504/rRNA_seqs/rRNA_hsap.fasta

Actually, looking at your IDs, they are unfortunately longer than 15 characters. This causes a problem for the Clustal tool which makes the alignments between different transcripts in the Ribopop pipeline. Because of this I would recommend providing a fasta file with your rRNA sequences and then manually truncating the IDs in that file (i.e. ENSRNAT049964327 -> RNAT049964327). Unfortunately, Ensembl BioMart doesn't seem to have this species, so I can't see a simple web-based way to extract all the rRNA sequences to a new file.

I can grep them out:

grep rRNA Coccidioides_immitis_rmscc_2394_gca_000149895.ASM14989v1.ncrna.fa | cut -f 1 -d ' '| cut -c 2- > Cimmitis_rRNA_ids.txt

Then use quick Biopython script to extract the sequences to a new file while removing the first 3 characters from the IDs:

python extract_rrna_cimmitis.py

Then you can decide which ones of these you want to include in the targets.csv file and list those there and also provide the path to this fasta file under 'file'. Does that make sense?
[Cimmitis_rRNA_ids.txt](https://github.com/marykthompson/ribopop_probe_design/files/6093308/Cimmitis_rRNA_ids.txt)

This file should be .fa extension, but GitHub doesn't accept that, so you can just change the extension to .fa
[Cimmitis_rRNA.txt](https://github.com/marykthompson/ribopop_probe_design/files/6093309/Cimmitis_rRNA.txt)

This is the python script, extension should be .py
[extract_rrna_cimmitis.txt](https://github.com/marykthompson/ribopop_probe_design/files/6093324/extract_rrna_cimmitis.txt)

**Comment by AMElmore** [comment]
Yes, that makes sense. Thank you so much for putting the Cimmitis_rRNA.fa file together, that is a huge help. I chose the targets and edited them into the targets.csv and params.cvs files, and the pipeline seems to be getting closer to finishing correctly! I'm still getting this error though, and I'm not sure how to fix it:

Error in rule download_seqs:
    jobid: 8
    output: offtarget_filtering/Cimmitis/ncrna.fa, offtarget_filtering/Cimmitis/cdna.fa, offtarget_filtering/Cimmitis/ann.gtf, offtarget_filtering/Cimmitis/genome.fa

RuleException:
CalledProcessError in line 75 of /Users/alexiselmore/Desktop/Cimmitis_probes/Snakefile:
Command 'set -euo pipefail;  /Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/bin/python3.8 /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmpgjtlo_ul.download_seqs.py' returned non-zero exit status 1.
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/Users/alexiselmore/Desktop/Cimmitis_probes/Snakefile", line 75, in __rule_download_seqs
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper

**Comment by AMElmore** [comment]
[design_params.zip](https://github.com/marykthompson/ribopop_probe_design/files/6095745/design_params.zip)

**Comment by marykthompson** [comment]
After looking at your files, I found the issue to be that the organism name does not match between your design param files.

targets.csv = 'Cimmitis'
seqs_and_anns.csv = 'Coccidioides_immitis'

Once I changed these both to 'Cimmitis', the pipeline was able to run. 

However, it did not choose any probes for the 5S, because there weren't any places to design probes not overlapping variant regions between different transcripts.

This is the consensus sequence of the 5S transcripts that you input and you can see that there are no ~30mers not covered by Ns (indicating a mismatch between transcripts).

>5S
NaNatacaacagcNNggattcgctggtNgtNaNccacccaactactagccNgccggcNNg
tNNNttNagtaNggctgagcggacgggaagccctNtNNtNNNcNNcctNtggtcgtatgt

Usually this is is not a problem for longer transcripts like the large and small subunit rRNAs, because there are fewer mismatches and more space to design the probes. I suggest that you go through the alignment file of the 5S and choose one or two representative transcripts to use for probe design and then submit them separately to the probe designer (they will then have to have unique names, like 5S_a and 5S_b). Alternatively, you could use a lower % consensus sequence as the target, I put the 5S.clustal file into MView and it generates a 70% consensus, for example.


[5S_variants.zip](https://github.com/marykthompson/ribopop_probe_design/files/6100194/5S_variants.zip)
https://www.ebi.ac.uk/Tools/msa/mview/

Are you only designing probes against the small rRNA transcripts? Not also the larger ones?

**Comment by AMElmore** [comment]
Ah I see! I went through and chose representative 5S sequences, as well as added representative LSU and SSU sequences. I had meant to add in the larger rRNA transcripts, I think I had just misunderstood when not including them. 

The pipeline is almost running normally, now I'm just getting this:

Traceback (most recent call last):
  File "/Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmpadrpmqx7.extract_target_seqs.py", line 48, in <module>
    main(sys.argv[1:])
  File "/Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmpadrpmqx7.extract_target_seqs.py", line 41, in main
    records.append(collect_seqs_by_id(i, ids))
  File "/Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmpadrpmqx7.extract_target_seqs.py", line 16, in collect_seqs_by_id
    with open(infasta, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/User/alexiselmore/Desktop/Cimmitis_probes/Cimmitis_rRNA.fa'
[Mon Mar  8 12:27:56 2021]
Error in rule extract_targets:
    jobid: 27
    output: target_sequences/original/Cimmitis/5S_b.fa

RuleException:
CalledProcessError in line 28 of /Users/alexiselmore/Desktop/Cimmitis_probes/Snakefile:
Command 'set -euo pipefail;  /Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/bin/python3.8 /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmpadrpmqx7.extract_target_seqs.py' returned non-zero exit status 1.
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/Users/alexiselmore/Desktop/Cimmitis_probes/Snakefile", line 28, in __rule_extract_targets
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/log/2021-03-08T122700.975351.snakemake.log

**Comment by AMElmore** [comment]
[Cimmitis_probes.zip](https://github.com/marykthompson/ribopop_probe_design/files/6103721/Cimmitis_probes.zip)

**Comment by marykthompson** [comment]
The problem is in the error message 
FileNotFoundError: [Errno 2] No such file or directory: '/User/alexiselmore/Desktop/Cimmitis_probes/Cimmitis_rRNA.fa'

The name of the file in your zip archive is Cimmitis_rRNA.fa.txt. The name in the targets.csv file is Cimmitis_rRNA.fa. You need to make sure that the name matches the file that you have.

Once I changed the name, the pipeline ran for me on a test set of 2 probes per target. You have to watch out for this error, it will say "Not enough space". This is what happened to me when I ran it with your original params.csv file that wanted to design 5 probes for the 5S and the 8S. These are very small, so you can't fit 5 probes there. I would try maybe 2 for those. The other ones are longer, you'll have to decide how many you want. I did every 400 nt for my probe set, you could go with every 200 or 300 nt if you want denser coverage.

I also discovered a small naming error which didn't effect the results, but was changing the name of 5S_a to 5S_ in the results table, so I have fixed that now in the code revision I just posted.

**Comment by marykthompson** [comment]
Also, I should mention that it's not necessary to choose a representative transcript for all your targets, it's just that the 5S was so variable that you needed to subset them to avoid the whole target space being covered by Ns. Ribopop will build a consensus sequence of your input transcripts by target and design probes not overlapping the Ns, so this may be the better approach for largely conserved transcripts with a few differences here and there.

**Comment by AMElmore** [comment]
Ah I missed that, thank you! I converted the .txt file to .fa with an online converter. I also took your advice and changed the number of probes from 5 to 2 for 5S_a, 5S_b, and 8s. The pipeline is so close to running now! I'm still getting an error though but it seems to be towards the end of the pipeline: 

Error in rule extract_targets:
    jobid: 27
    output: target_sequences/original/Cimmitis/5S_b.fa

RuleException:
CalledProcessError in line 28 of /Users/alexiselmore/Desktop/Cimmitis_probes/Snakefile:
Command 'set -euo pipefail;  /Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/bin/python3.8 /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/scripts/tmp5k7a01sq.extract_target_seqs.py' returned non-zero exit status 1.
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/Users/alexiselmore/Desktop/Cimmitis_probes/Snakefile", line 28, in __rule_extract_targets
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/Users/alexiselmore/opt/miniconda3/envs/ribopop_probe_design/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/log/2021-03-09T130623.632989.snakemake.log

**Comment by marykthompson** [comment]
It ran to completion for me with the last version of the files you sent after I changed the name of the fasta file and decreased the number of probes to 2.

extract_targets.py is actually quite near the beginning of the pipeline and it is certainly a naming issue like the ones I previously described. If you want to send your new files I can take a look. Otherwise, please go through and make sure all the names of the files and targets match across all your input files. The error occurred trying to make the 5S_b.fa file, so you should focus on that one. Does the ID of the transcript you provided for 5S_b match a fasta entry in your input fasta file?

You can try seeing what's in that log with 'cat /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/log/2021-03-09T130623.632989.snakemake.log'. It may have some more specific information about what caused the error.

**Comment by AMElmore** [comment]
I ran 'cat /Users/alexiselmore/Desktop/Cimmitis_probes/.snakemake/log/2021-03-09T130623.632989.snakemake.log' to see if I could tell any more about the problem, and didn't see much to add to the error besides: 

[Tue Mar  9 13:07:16 2021]
rule extract_targets:
    input: offtarget_filtering/Cimmitis/ncrna.fa
    output: target_sequences/original/Cimmitis/5S_b.fa
    jobid: 27
    wildcards: org=Cimmitis, target=5S_b

[Tue Mar  9 13:07:18 2021]
Error in rule extract_targets:
    jobid: 27
    output: target_sequences/original/Cimmitis/5S_b.fa

I tried deleting 5S_b and changed 5S_a just to 5S to hopefully eliminate the problem, but then it began having problems with the output for LSU. 

Error in rule extract_targets:
    jobid: 27
    output: target_sequences/original/Cimmitis/LSU.fa

I'm not sure where the naming error is happening, I've searched params.csv and targets.csv but everything seems to line up. Here is the file I'm using
[Cimmitis_probes.zip](https://github.com/marykthompson/ribopop_probe_design/files/6124282/Cimmitis_probes.zip)

**Comment by marykthompson** [comment]
Hi, sorry, but I don't know. Those files work fine for me now. Maybe try creating a fresh directory and starting with only the design_params/ folder and config.yml then rerun? It's possible that something from the failed Snakemake runs in the old directory is causing problems. It also worked fine for me using the 5S_a and 5S_b versions that you posted previously after I fixed the fasta file naming error.

If you want, you can post the params you want here (with 5S_a and 5S_b) and I can run it for you and put the results here.

**Comment by AMElmore** [comment]
I tried re-downloading the pipeline and making a new directory, but I get the same error. If you ran it and posted and results I would appreciate it immensely, thank you so much. Here is the current file with 5_a and 5_b params. 
[Cimmitis_probes.zip](https://github.com/marykthompson/ribopop_probe_design/files/6135013/Cimmitis_probes.zip)

**Comment by marykthompson** [comment]
Hi, here is the output of the pipeline. I made two versions, set 1, with 2 probes each for 5S_a, 5S_b, 8S, and 5 probes each for LSU and SSU. I thought the coverage seemed a little low for the length of LSU and SSU, so I also took the liberty of making a second set, set2, with 7 probes for SSU and 9 probes for LSU, in case you want to go with that.

I also changed the Tm_window_size param to approximately target length/ (num desired probes * 2) which I found can lead to more evenly chosen probes.

If you look at the output for 5S_a and 5S_b, the first probes selected are very similar (identical except for a 1 nt offset shift), so I would choose only one of those to avoid duplication in your set. The second ones chosen are quite different, reflecting sequence differences between 5S_a and 5S_b, so I would keep both of those.

[Cimmitis_set1_16pr.zip](https://github.com/marykthompson/ribopop_probe_design/files/6148130/Cimmitis_set1_16pr.zip)
[Cimmitis_set2_22pr.zip](https://github.com/marykthompson/ribopop_probe_design/files/6148135/Cimmitis_set2_22pr.zip)
